Initial implementation of supervisor client

mdegat01 opened the pull request:
Most of initial setup of the repo including project files, ci setup as well as initial implementation of the client for addons only.

MartinHjelmare commented on aiosupervisor/client.py:
```diff
@@ -0,0 +1,203 @@
+        *,
+        params: dict[str, str] | None = None,
+        response_type: ResponseType = ResponseType.JSON,
+    ) -> Awaitable[Response]:
```
Don't return awaitables. Make the methods coroutine functions and await `self._request`. We don't need to avoid coroutine chaining.

mdegat01 replied:
Ah ok. Supervisor did that a lot, I assumed it was for performance reasons. If its not a big deal I'll change it here and maybe change that behavior in supervisor over time. Thanks!

MartinHjelmare replied:
It's yielding to the event loop that is expensive. Just awaiting a coroutine function isn't expensive. I think this was a misunderstanding from some people in the team in the early days.

MartinHjelmare commented on aiosupervisor/client.py:
```diff
@@ -0,0 +1,203 @@
+    if "application/json" not in content_type:
+        if raise_on_fail:
+            raise SupervisorError(
+                f"Unexpected response received from supervisor when expecting JSON. Status: {response.status}, content type: {content_type}"
```
It's nice to break long strings around max 88 characters per line.

MartinHjelmare commented on README.md:
```diff
@@ -1,2 +1,86 @@
+
+loop = asyncio.new_event_loop()
+asyncio.set_event_loop(loop)
+loop.run_until_complete(supervisor_client.info())
```
Use `asyncio.run`.

mdegat01 commented on pyproject.toml:
```diff
@@ -0,0 +1,370 @@
+[tool.pylint."MESSAGES CONTROL"]
+# Reasons disabled:
+# format - handled by ruff
+# abstract-method - with intro of async there are always methods missing
```
Yea all this was copied over from core's pyproject.toml. I'll make the changes you suggested here and below, I had kind of assumed we wanted to standardize our linting in Home Assistant projects unless there was good reason to deviate though. But if these don't make sense here I'll remove them, I'll admit I didn't look into each closely.

mdegat01 commented on .github/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
I was using python-mealie as a reference for a lot of this and saw that it did that here: <https://github.com/joostlek/python-mealie/blob/main/.github/LICENSE.md> and here <https://github.com/joostlek/python-mealie/blob/main/LICENSE.md>. I did the same but meant to ask him about why it was that way. Seems like it was a mistake, will remove.

mdegat01 commented on setup.py:
```diff
@@ -0,0 +1,11 @@
+REQUIREMENTS_FILE = SUPERVISOR_CLIENT_DIR / "requirements.txt"
+REQUIREMENTS = REQUIREMENTS_FILE.read_text(encoding="utf-8")
+
+setup(dependencies=REQUIREMENTS.split("/n"))
```
Is that something that should be changed here? Or are you saying in the requirements file I should switch most (if not all) of the `==` to `>=` for version specifiers? And at least remove any patch versions when present as those are unlikely to cause a breakage

Or are you saying try to remove version specification entirely from `requirements.txt` until we find there is an issue with a particular version?

MartinHjelmare replied:
When using requirements files in a library together with setup.py I use the requirement files for development only and duplicate the requirements directly in setup.py without pinning as far as possible. The requirements in the requirement files are pinned strictly.

When using a package tool that keeps track of installed package versions, like poetry, you don't need requirement files and you get exact versions installed for development and forgiving versions for production.

Only using a pyproject.toml file without requirement files and a build backend like setuptools requires duplicating the requirements in the dependencies and test dependencies, I think, to achieve it, pinning the test dependencies strictly.

mdegat01 commented:
@MartinHjelmare Ok I believe I implemented everything from your suggestions.

@agners In addition I made the changes we discussed to allow supervisor to add values to some enums in future as a non-breaking change for client. I updated the docstring of such classes to mention this and then ensured the places they were used in the models included `| str` to support parsing unknown values. Ideally clients would then just write a case switch on the enum values and ignore everything else but I guess I'll leave it up to them. It seems better to return the string then to try and strip it out. Let me know if you disagree with any of my choices of which enums I said allowed non-breaking changes and which I didn't.

MartinHjelmare commented on aiosupervisor/models/addons.py:
```diff
@@ -13,7 +13,11 @@
+    """AddonStage type.
+
+    Supervisor may add new AddonStages in future as a non-breaking change.
+    Those should be returned as string in response models until next update.
```
Doesn't this mean that clients can't trust this model?

agners replied:
We'd only would add new stages. So you can trust that the currently existing three will be there.

It depends a bit on the use case of a string enum if a library user can make safe assumptions/implement it in a forward compatible way. E.g. the `HostFeature` enum is quite straight forward. New features will be added, but a library user should just not care.

This is definitely a bit more controversial: Currently we have `experimental`, `stable`, `deprecated`, and the reflect the life cycle in that order. If we'd introduce `beta` which presumably is between `experimental` and `stable`, a client library would most likely handle this wrong :cry: 

For the library user, it probably would be better if commit to the current stages and keep this one fixed :thinking: .

We do have add-on channel on the road map, so that ties into stages. These will be `stable`, `beta` and `development`, and will most likely more important then the add-on stage really.

IMHO we should stick with the three stages we have, so make fields using `AddonStage` just taking `AddonStage`.

mdegat01 replied:
No that wasn't the intention. The situation we're dealing with here is that Supervisor wants to be able to add new values to these some of fields with a fixed set of values in the future. But the problem is this:
1. Supervisor auto-updates
2. Supervisor must be updated before everything else (core, addons, etc.)

That means these enums basically can never change without breaking all existing clients, not even additive changes (it's ok that removing a value breaks clients as Supervisor should never do that).

When talking it over with Stefan what we settled on was for enums supervisor wants to be able to make additive changes to we note that somewhere. And then when a field uses that enum the type of the field is `AddonStage | str`. That way if Supervisor adds a new `AddonStage` existing clients won't break, it'll just be a string.

We do not expect users to try and handle string values though. They should just write a case switch based on the enum. And then the next time the supervisor client updates itll incorporate that new value into the enum and they can update their case case switch.

mdegat01 replied:
> Adding new add-on stages is definitely a bit more controversial: Currently we have experimental, stable, deprecated, and the reflect the life cycle in that order. If we'd introduce something new, e.g. say beta, which presumably would be between experimental and stable, a client library would most likely handle this wrong 😢

I'm not understanding why this one would be controversial. To my knowledge no current client of supervisor uses this information anywhere (including core and frontend). It's effectively just metadata we have available that no one really adopted. Making it a breaking change requiring a new version of the API to add a new stage seems very heavy handed.

What kind of use cases are you thinking of @agners where someone would use this where a new stage would break them?

MartinHjelmare commented on pyproject.toml:
```diff
@@ -0,0 +1,369 @@
+]
+
+[tool.ruff]
+lint.select = [
```
I'd select all instead and then ignore the ones we don't like.

MartinHjelmare commented on pyproject.toml:
```diff
@@ -0,0 +1,369 @@
+]
+
+[tool.ruff.lint.flake8-import-conventions.extend-aliases]
+voluptuous = "vol"
```
We don't use voluptuous.

MartinHjelmare commented on pyproject.toml:
```diff
@@ -0,0 +1,369 @@
+[tool.setuptools.packages.find]
+include = ["aiosupervisor*"]
+
+[tool.pylint.MAIN]
```
I'd consider dropping pylint completely. Ruff does a lot already and we don't use any custom pylint plugins here.

MartinHjelmare commented on requirements_dev.txt:
```diff
@@ -0,0 +1,23 @@
+# Production requiremetns
```
```suggestion
# Production requirements
```

MartinHjelmare commented on pyproject.toml:
```diff
@@ -0,0 +1,369 @@
+]
+keywords = ["docker", "home-assistant", "api", "client-library"]
+requires-python = ">=3.12.0"
+dependencies = [
```
There's the `project.optional-dependencies` section we could add and have a `dev` list with the dev dependencies that are all pinned.

Then we don't need `requirements_dev.txt`, I think.

mdegat01 replied:
Does that mean I would need to introduce a toml parser here:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/.github/workflows/ci.yaml#L50

And here:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/script/devcontainer_bootstrap#L6

To set up environments in those places? Or is there a way to tell pip to install it as dev and set up the dependencies listed there accordingly?

MartinHjelmare replied:
Yes, each list name is an "extra" set of dependencies and can be installed with eg `pip install somelibrary[extraname]`.

mdegat01 replied:
@MartinHjelmare I had borrowed the venv caching from supervisor throughout the github actions. But how it works is by including a hash of the requirements as a key as you can see here:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/.github/workflows/ci.yaml#L41-L43

Is there a good way to replicate this behavior without the separate requirements file? Perhaps just hash the `pyproject.toml` file itself?

MartinHjelmare replied:
Yes, use that file.

agners commented on aiosupervisor/models/addons.py:
```diff
@@ -0,0 +1,342 @@
+):
+    """InstalledAddonComplete model."""
+
+    state: AddonState
```
So this restricts us to the currently existing add-on state. We did add `startup` relatively recently with [Supervisor PR #4376](https://github.com/home-assistant/supervisor/pull/4376/files#diff-dc1d068b16705906f72e7f53c871400e85e8695d332bde81fb2cf01eec62cf7e). 

IMHO, introducing new states is very on the edge for a "backwards compatible" change :sweat_smile: . So I prefer to not keep this open.

BUT: Looking at the states, it seems the only one which might be missing is `stopping`. Should we maybe just introduce that one and then keep them fixed? :thinking:

mdegat01 replied:
I guess my thinking was that existing clients (frontend and core mostly) already closely watch and depend on this field. So changing the values here already felt like a breaking change.

I don't know though. I don't really have a good answer here. Supervisor auto-updates so I'm pretty tempted to just put `| str` on all of them just in case, I just don't like the model that makes.

mdegat01 replied:
I guess I'll just note on here that regardless of whether we change this value in the future or not this one seems like one that it should actually be breaking to add a value. Clients can and will expect to know the state of addons. If we add a new value in the future then they will be clueless as to what state it is in.

Out of all the enums the ones for State of Supervisor and Addons seem like the most obvious candidate for even additive changes are breaking. The others I'm less sure about.

mdegat01 commented on aiosupervisor/client.py:
```diff
@@ -0,0 +1,205 @@
+    """Check if response is json according to Content-Type."""
+    content_type = response.headers.get("Content-Type", "")
+    if "application/json" not in content_type:
+        if raise_on_fail:
+            raise SupervisorError(
+                "Unexpected response received from supervisor when expecting"
+                f"JSON. Status: {response.status}, content type: {content_type}"
+            )
```
It's used in two places. First one is in error handling here:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L115

In this case I don't want it to raise if its not JSON because I have already set an exception type based on the status code. I want to raise that instead to give the client as much info as I can.

The second is if we got a successful response from supervisor but it wasn't JSON when we expected JSON:
https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L122

In this case raising a generic `SupervisorError` is fine because this is already an unknown error situation. I choose to raise from the method as opposed to returning false and then raising because it felt cleaner to have the error message in one place (in case I used the function again) and I wouldn't have to duplicate the code to get the content type header.

mdegat01 replied:
I didn't make any changes here, are we good with the way this works?

agners replied:
> https://github.com/home-assistant-libs/python-supervisor-client/blob/b0eb227081859b4e5d2c6fe06d8e444cc02097a5/aiosupervisor/client.py#L122
> 

IMHO, `is_json` should not have such side effects. `is_json` for me indicates a function which is very passive.


> I choose to raise from the method as opposed to returning false and then raising because it felt cleaner to have the error message in one place (in case I used the function again) and I wouldn't have to duplicate the code to get the content type header.

Yeah that is a very valid concern, and that is why I suggested two separate functions. This way we don't duplicate code, while still have function names which clearly state what the function is doing.

Anyhow, this is more stylistic, and internal anyways. I also don't care that much, we can leave as is if you prefer.

mdegat01 commented on pyproject.toml:
```diff
@@ -0,0 +1,369 @@
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "aiosupervisor"
```
`aiohasupervisor` is kind of a mouthful. And `home-assistant-supervisor-client` seems way too long. I guess I just figured the doc/description would take care of explaining what specific supervisor this was about?

If I had to pick between them I guess I'd probably pick `aiohasupervisor` over the much longer one though.

MartinHjelmare replied:
`aiohasupervisor` sounds ok.

mdegat01 commented:
Ok @MartinHjelmare and @agners I made the changes requested above. Linting is re-written to basically just run ruff and no more pylint, project is renamed to aiohasupervisor and all the tweaks are handled.

On the biggest topic of discussion, the enums, here's what I settled on:
1. I removed the caveat from nearly all of them. In nearly all cases adding a new value to those fields will be a breaking change
2. I made two exceptions: `HostFeature` and `Capability`. For both of these cases the enum is not trustworthy as a model, users should treat it like a subset of the options. Thats because for both of these Supervisor adds support for new host features and privileged capabilities regularly. And the fields that use these are basically just metadata used for debugging purposes (if a host feature is missing that's possibly the root cause, etc.) I don't expect hard logic based on these but if someone wants to write that they'll just have to be aware there may be more then what is listed.
3. I think in general I'll proceed with the idea that most of these enums in Supervisor are going to be strongly typed parts of the model unless there's a good reason for an exception like above. We'll have to look into how best to version Supervisor's API going forward.

agners commented:
> On the biggest topic of discussion, the enums, here's what I settled on:
> 
>     1. I removed the caveat from nearly all of them. In nearly all cases adding a new value to those fields will be a breaking change
> 
>     2. I made two exceptions: `HostFeature` and `Capability`. For both of these cases the enum is not trustworthy as a model, users should treat it like a subset of the options. Thats because for both of these Supervisor adds support for new host features and privileged capabilities regularly. And the fields that use these are basically just metadata used for debugging purposes (if a host feature is missing that's could be the root cause of the issue, etc.) I don't expect hard logic based on these but if someone wants to write that they'll just have to be aware there may be more then what is listed.
> 

First I'd like to note that as long as we don't make an stable client library release, we can discuss and change policy on this particular issues. So I'd like to *not* hold back merging this PR over this particular discussion.

That said, in general, we had a quite loose policy when it comes to API changes so far: We only operate on a single API version essentially. Afaik, basically whatever keeps old Core version generally working was free game. Things which add-on potentially used we were a bit more strict too. But other than that, changes, especially extensions, have been made without versioning.

Technically, introducing a client library and having a new policy on API changes are two different issues. That said, with mashumaro as library we automatically make things more strict, so in a way the introduction of this client library *will* have impact on what flexibility we have in the future.

The suggested middle ground works for me. It is a step into being more strict, while still leaving the door open for things we are quite certain will change and don't have breaking impact typically. Ideally we should also add a note in the API docs. 

The only enum I am bit wary currently is `AddonState`. Maybe we should sneak in a `stopping` state (just as we did with `startup` a while ago), and then call it fixed? :thinking: 

>     3. I think in general I'll proceed with the idea that most of these enums in Supervisor are going to be strongly typed parts of the model unless there's a good reason for an exception like above. We'll have to look into how best to version Supervisor's API going forward.

Typically, there is a v1/v2 type of endpoint. There is no `v1` currently, but we could just treat everything which has no version in the path as v1, and start with `/v2/...`. The problem is a bit the way we work is quite iterative. So would we just increment on every API change? That will bump the version quite often... Or will we collect a couple of API changes and then do a big release? That doesn't fit very well with our development model :cry:

Fatal1ty commented:
@mdegat01 As an alternative option, you can have an Enum forward compatible using the [`_missing_`](https://docs.python.org/3/library/enum.html#enum.Enum._missing_) method:

```python
from enum import Enum

class MyEnum(Enum):
    FOO = "foo"
    BAR = "bar"
    UNKNOWN = "unknown"

    @classmethod
    def _missing_(cls, value):
        return cls.UNKNOWN

assert MyEnum("baz") == MyEnum.UNKNOWN
```

mdegat01 commented:
@Fatal1ty huh, I didn't know that. That's what I was looking for earlier haha.

I think for the two cases discussed above I'd prefer to use `| str` though. For those two I expect the primary use case to be simply displaying what features/capabilities are enabled to the user for debugging purposes. So by doing `| str` and having mashumaro keep the value as a string if its not the enum even older clients basically work fine with unknown values. Whereas if I mapped those to `UNKNOWN` the information would be lost in the display.

However @agners do we want to use this for any other enums? I guess I'd kind of warmed up to the idea of a strongly typed model for the other enums and versioning supervisor. But this is how you implement the alternative we had been discussiong for some of them, mapping an unknown value to `UNKNOWN`.

MartinHjelmare commented on pyproject.toml:
```diff
@@ -20,6 +20,31 @@ dependencies = [
+    "mypy==1.10.1",
+    "pre-commit==3.7.1",
+    "pytest-aiohttp==1.0.5",
+    "pytest-asyncio==0.23.6",
```
I don't think we need this since pytest-aiohttp depends on it.
```suggestion
```

MartinHjelmare commented on pyproject.toml:
```diff
@@ -20,6 +20,31 @@ dependencies = [
+    "pytest==8.2.2",
+    "ruff==0.5.2",
+    "time-machine==2.14.2",
+    "urllib3==2.2.2",
```
What do we use urllib3 for?

MartinHjelmare commented on pyproject.toml:
```diff
@@ -0,0 +1,127 @@
+    "pytest-timeout==2.3.1",
+    "pytest==8.2.2",
+    "ruff==0.5.2",
+    "time-machine==2.14.2",
```
Where is time-machine used?

MartinHjelmare commented on requirements_dev.txt:
```diff
@@ -0,0 +1,21 @@
+# Production requirements
```
Can we remove this file?

MartinHjelmare commented on tests/conftest.py:
```diff
@@ -0,0 +1,27 @@
+    async with SupervisorClient(
+        SUPERVISOR_URL,
+        "abc123",
+    ) as mealie_client:
```
```suggestion
    ) as supervisor_client:
```

MartinHjelmare commented on tests/conftest.py:
```diff
@@ -0,0 +1,27 @@
+        SUPERVISOR_URL,
+        "abc123",
+    ) as mealie_client:
+        yield mealie_client
```
```suggestion
        yield supervisor_client
```

agners commented:
> However @agners do we want to use this for any other enums? I guess I'd kind of warmed up to the idea of a strongly typed model for the other enums and versioning supervisor's API more. But this is how you implement the alternative we had been discussing for some of them, mapping an unknown value to `UNKNOWN`.

Yeah in the end it is a different implementation, instead of allowing `str` type simply create Unkown as a Enum type. It feels cleaner as it is "stronger" typed. But it has the downside that the API user would not be able to "see" what underlying string is not yet supported. But IMHO, not a big deal, it would at most be informational.

So, either way works for me, leave as is or implement unknown (for the ones we use `| str` right now).

mdegat01 commented:
> The only enum I am bit wary currently is AddonState. Maybe we should sneak in a stopping state (just as we did with startup a while ago), and then call it fixed? 🤔

I mean we can but it doesn't seem ideal to me to have the model list an option that isn't used. I also don't think we have any plans to do a `stopping` state? We did `starting` because of the frontend but `stopping` isn't one we needed I don't think.

agners commented:
> I also don't think we have any plans to do a `stopping` state? We did `starting` because of the frontend but `stopping` isn't one we needed I don't think.

No plans currently. But in theory, between issuing a command to stop a container, and the container actually being stopped can take some time. And we don't handle this with a separate state. It might make sense to know that the container is being stopped right now. Just saying, that is the only potential state which we might miss sometime in the feature.

MartinHjelmare reviewed (approved):
I think this is ok, but I'd recomend using the `_missing_` method in the enums where needed.

agners commented:
In discussion with @mdegat01 another solution would be to make the `privileged` field (type list of `Capability`) and `features` field (type list of `HostFeature`) of type `str` :man_shrugging: . But this to me seems a loss in typing, and is worse over `| str`. So I think for these two fields the current solution is fine.

agners commented on aiohasupervisor/models/root.py:
```diff
@@ -0,0 +1,117 @@
+class AvailableUpdate(ResponseData):
+    """AvailableUpdate type."""
+
+    update_type: UpdateType | str
```
Let's remove `| str` here, there is no new update type on the horizon.
